MantidQt/API/src/NonOrthogonal.cpp

```diff
@@ -0,0 +1,227 @@
+#include "MantidQtAPI/NonOrthogonal.h"
+#include "MantidAPI/IMDWorkspace.h"
+#include "MantidAPI/Sample.h"
+#include "MantidAPI/Run.h"
+#include "MantidAPI/CoordTransform.h"
+#include "MantidKernel/Matrix.h"
+#include "MantidGeometry/Crystal/UnitCell.h"
+#include "MantidGeometry/Crystal/OrientedLattice.h"
+#include "MantidAPI/IMDEventWorkspace.h"
+#include "MantidAPI/IMDHistoWorkspace.h"
+#include "MantidGeometry/MDGeometry/HKL.h"
+
+#include <boost/pointer_cast.hpp>
+
+namespace {
+void checkForSampleAndRunEntries(
+    const Mantid::API::Sample &sample, const Mantid::API::Run &run,
+    Mantid::Kernel::SpecialCoordinateSystem specialCoordinateSystem) {
+
+  if (Mantid::Kernel::HKL != specialCoordinateSystem) {
+    throw std::invalid_argument(
+        "Cannot create non-orthogonal view for non-HKL coordinates");
+  }
+
+  if (!sample.hasOrientedLattice()) {
+    throw std::invalid_argument("OrientedLattice is not present on workspace");
+  }
+
+  if (!run.hasProperty("W_MATRIX")) {
+    throw std::invalid_argument("W_MATRIX is not present on workspace");
+  }
+}
+
+void normalizeColumns(Mantid::Kernel::DblMatrix &skewMatrix) {
+  const auto numberOfColumns = skewMatrix.numCols();
+  const auto numberOfRows = skewMatrix.numRows();
+  std::vector<double> bNorm;
+  bNorm.reserve(skewMatrix.numCols());
+  for (size_t column = 0; column < numberOfColumns; ++column) {
+    double sumOverRow(0.0);
+    for (size_t row = 0; row < numberOfRows; ++row) {
+      sumOverRow += std::pow(skewMatrix[row][column], 2);
+    }
+    bNorm.push_back(std::sqrt(sumOverRow));
+  }
+
+  // Apply column normalisation to skew matrix --> TODO: Check why 3 is
+  // hardcoded
+  Mantid::Kernel::DblMatrix scaleMat(3, 3, true);
+  scaleMat[0][0] /= bNorm[0];
+  scaleMat[1][1] /= bNorm[1];
+  scaleMat[2][2] /= bNorm[2];
```

Comment:
Would it make sense to have something like:
```cpp
const size_t dim = 3;
Mantid::Kernel::DblMatrix scaleMat(dim , dim , true);
for (size_t index = 0; index < dim; ++index) {
  scaleMat[dim][dim] /= bNorm[dim];
}
```

We should follow up on that TODO or remove it.